File: SSSystemGenerator/SSSystemGenerator/Classes/SavingProccess/ImageWorks.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DataEmbeddingInImageTest
{
    public class ImageWorks
    { // https://github.com/ktekeli/img-stego

        private byte getByte(byte[] bits)
        {
            String bitString = "";
            for (int i = 0; i < 8; i++)
                bitString += bits[i];
            byte newpix = Convert.ToByte(bitString, 2);
            int dePix = (int)newpix;
            return (byte)dePix;
        }

        private byte[] getStringBits(string data)
        {
            byte[] text = System.Text.ASCIIEncoding.ASCII.GetBytes(data);
            BitArray bits = new BitArray(text);
            bool[] boolarray = new bool[bits.Count];
            bits.CopyTo(boolarray, 0);
            byte[] bitsArray = boolarray.Select(bit => (byte)(bit ? 1 : 0)).ToArray();
            Array.Reverse(bitsArray);
            return bitsArray;
        }

        private byte[] getBits(byte simplepixel)
        {
            int pixel = 0;
            pixel = (int)simplepixel;
            BitArray bits = new BitArray(new byte[] { (byte)pixel });
            bool[] boolarray = new bool[bits.Count];
            bits.CopyTo(boolarray, 0);
            byte[] bitsArray = boolarray.Select(bit => (byte)(bit ? 1 : 0)).ToArray();
            Array.Reverse(bitsArray);
            return bitsArray;
        }

        /*This method is used to embed data into pixels*/
        public Color embed(Color pixel, byte[] bits)
        {

            byte[] RedBits = getBits((byte)pixel.R);
            byte[] GreenBits = getBits((byte)pixel.G);
            byte[] BlueBits = getBits((byte)pixel.B);

            /*LSB substition of RGB values is done as following:
            Red: Last 3 bits, Green: Last 3 bits, Blue: Last 2 bits
            This process lets us embed 1 byte in each pixel*/

            #region BitChange
            RedBits[5] = bits[0];
            GreenBits[5] = bits[1];
            RedBits[6] = bits[2];
            RedBits[7] = bits[3];
            GreenBits[6] = bits[4];
            GreenBits[7] = bits[5];
            BlueBits[6] = bits[6];
            BlueBits[7] = bits[7];

            byte newRed = getByte(RedBits);
            byte newGreen = getByte(GreenBits);
            byte newBlue = getByte(BlueBits);

            return Color.FromArgb(newRed, newGreen, newBlue);
            #endregion

        }

        /*This method is used to extract data from pixels*/
        public byte extract(Color pixel)
        {
            byte[] RedBits = getBits((byte)pixel.R);
            byte[] GreenBits = getBits((byte)pixel.G);
            byte[] BlueBits = getBits((byte)pixel.B);
            byte[] BitsToDecrypt = new byte[8];

            BitsToDecrypt[0] = RedBits[5];
            BitsToDecrypt[1] = GreenBits[5];
            BitsToDecrypt[2] = RedBits[6];
            BitsToDecrypt[3] = RedBits[7];
            BitsToDecrypt[4] = GreenBits[6];
            BitsToDecrypt[5] = GreenBits[7];
            BitsToDecrypt[6] = BlueBits[6];
            BitsToDecrypt[7] = BlueBits[7];

            return getByte(BitsToDecrypt);
        }

        public Bitmap TextToBitmap(string text, Bitmap sourceImage)
        {
            string textbox = text;
            char[] textarray = text.ToArray();

            Bitmap mainImage = new Bitmap(sourceImage);

            /* Encoding process */
            #region Encoding

            //Embed type of data into last 3 pixels.
            #region type_embed
            // "tt1" is the code to define hidden data is a text message. (type:text:1)
            string[] type = new string[] { "t", "t", "1" };

            for (int j = 0; j < 3; j++)
            {
                Color pixel = mainImage.GetPixel(mainImage.Width - j - 1, mainImage.Height - 1);
                pixel = embed(pixel, getStringBits(type[j]));
                mainImage.SetPixel(mainImage.Width - j - 1, mainImage.Height - 1, pixel);
            }

            #endregion

            // Embed length of message into 13 pixels in reverse [3:15]
            #region length_embed

            string a = Convert.ToString((text).Length);
            a = a.PadLeft(13, '0'); //Zero-padding
            char[] b = a.ToArray();

            for (int j = 3; j < 16; j++)
            {
                string aString = Convert.ToString(b[j - 3]);
                Color pixel = mainImage.GetPixel(mainImage.Width - j - 1, mainImage.Height - 1);
                pixel = embed(pixel, getStringBits(aString));
                mainImage.SetPixel(mainImage.Width - j - 1, mainImage.Height - 1, pixel);
            }

            #endregion

            int k = 0;

            for (int i = 0; i < mainImage.Height; i++)
            {
                for (int j = 0; j < mainImage.Width; j++)
                {
                    if (k < textbox.Length)
                    {
                        string msg = Convert.ToString(textarray[k]);
                        Color pixel = mainImage.GetPixel(j, i);
                        pixel = embed(pixel, getStringBits(msg));
                        mainImage.SetPixel(j, i, pixel);
                        k++;
                    }

                }
            }

            #endregion

            return mainImage;
        }

        public string BitmapToText(Bitmap encodedImage)
        {

            string outputText = "";

            /* Decoding process */
            #region Decoding

            int k = 0;
            string tlength = "";

            #region length_extract

            for (int j = 3; j < 16; j++)
            {
                Color pixelToDecode = encodedImage.GetPixel(encodedImage.Width - j - 1, encodedImage.Height - 1);
                byte delength = extract(pixelToDecode);
                tlength += Convert.ToInt32(Encoding.ASCII.GetString(BitConverter.GetBytes(delength)));
            }

            int length = Convert.ToInt32(tlength);

            #endregion

            k = 0;

            for (int i = 0; i < encodedImage.Height; i++)
            {
                for (int j = 0; j < encodedImage.Width; j++)
                {
                    if (k < length)
                    {
                        Color pixelToDecode = encodedImage.GetPixel(j, i);
                        byte demsg = extract(pixelToDecode);
                        string messageChar = Encoding.ASCII.GetString(BitConverter.GetBytes(demsg));
                        outputText += messageChar;
                        k++;
                    }
                }
            }

            #endregion

            return outputText.Replace("\0", "");
        }

    }
}

File: src/SGT/SGT/Parser/ImageWorks.java
package SGT.SGT.Parser;

import com.fs.starfarer.api.graphics.SpriteAPI;
import sun.security.util.ArrayUtil;

import java.awt.*;
import java.awt.image.BufferedImage;
import java.nio.ByteBuffer;
import java.nio.FloatBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Collections;

public class ImageWorks {
    // https://github.com/ktekeli/img-stego

    private byte getByte(byte[] bits) {
        String bitString = "";
        for (int i = 0; i < 8; i++)
            bitString += bits[i];
        byte newpix = Byte.parseByte(bitString, 2);
        int dePix = (int) newpix;
        return (byte) dePix;
    }

    private byte[] getBits(byte simplepixel) {
        String byteString = Integer.toBinaryString(simplepixel);

        while (true) {
            if (byteString.length() == 8)
                break;

            byteString = "0" + byteString;
        }

        byte[] bitsArray = new byte[8];

        char[] charArray = byteString.toCharArray();
        for (int i = 0; i < charArray.length; i++) {
            char bit = charArray[i];
            bitsArray[i] = (byte) Integer.parseInt(String.valueOf(bit));
        }

        return bitsArray;

//        Array.Reverse(bitsArray);
    }


    /*This method is used to extract data from pixels*/
    public byte extract(Color pixel) {

        byte[] RedBits = getBits((byte) pixel.getRed());
//        RedBits = reverseArray(RedBits);
        byte[] GreenBits = getBits((byte) pixel.getGreen());
//        GreenBits = reverseArray(GreenBits);
        byte[] BlueBits = getBits((byte) pixel.getBlue());
//        reverseArray(BlueBits);
        byte[] BitsToDecrypt = new byte[8];

        BitsToDecrypt[0] = RedBits[5];
        BitsToDecrypt[1] = GreenBits[5];
        BitsToDecrypt[2] = RedBits[6];
        BitsToDecrypt[3] = RedBits[7];
        BitsToDecrypt[4] = GreenBits[6];
        BitsToDecrypt[5] = GreenBits[7];
        BitsToDecrypt[6] = BlueBits[6];
        BitsToDecrypt[7] = BlueBits[7];

        return getByte(BitsToDecrypt);
    }

    public String BitmapToText(BufferedImage encodedImage, SpriteAPI texture) {


        String outputText = "";

        /* Decoding process */
        //region Decoding

        int k = 0;
        String tlength = "";

        //region length_extract

        for (int j = 3; j < 16; j++) {

            // y * width + x

            Color pixelToDecode = new Color((int) encodedImage.getRGB((int) (texture.getWidth() - j - 1), (int) (texture.getHeight() - 1)));
            byte delength = extract(pixelToDecode);

            String ourChar = new String(new byte[]{delength}, StandardCharsets.UTF_8);

            tlength += ourChar;
        }

        int length = Integer.parseInt(tlength.toString());

        //endregion

        k = 0;

        for (int i = 0; i < texture.getHeight(); i++) {
            for (int j = 0; j < texture.getWidth(); j++) {
                if (k < length) {
                    Color pixelToDecode = new Color((int) encodedImage.getRGB(j, i));
                    byte demsg = extract(pixelToDecode);
                    String messageChar = new String(new byte[]{demsg}, StandardCharsets.UTF_8);
                    outputText += messageChar;
                    k++;
                }
            }
        }

        //endregion

        return outputText.replace("\0", "");
    }

    private int coordinateToIndex(int x, int y, int width) {
        return y * width + x;
    }

    private byte[] reverseArray(byte[] array) {
        for (int i = 0; i < array.length / 2; i++) {
            byte temp = array[i];
            array[i] = array[array.length - i - 1];
            array[array.length - i - 1] = temp;
        }

        return array;
    }
}
